File: packages/VBCore/Sources/VBCore/Court.swift
// The court, as it is actually arranged on the floor.
//
// This is what the watch draws and what the phone's picker draws, from one definition, so
// the two cannot disagree about who serves next. Derived on read like every other figure:
// a stored court would be one more thing that can fall out of step with the serves.
//
//     4     3     2  <- on deck: rotates into service next
//     5     6     1  <- serving
//
// The arrangement stays still and the players move through it. Rotation is clockwise: the
// server steps to the bottom-middle position, and the top-right player steps down into
// service.
import Foundation

/// A position on the court, numbered as volleyball numbers them.
public enum CourtPosition: Int, CaseIterable, Sendable {
    case service = 1
    case rightFront = 2
    case middleFront = 3
    case leftFront = 4
    case leftBack = 5
    case middleBack = 6

    /// How far along the serving order this position stands from whoever is serving.
    ///
    /// The service position is the server; the player at position 2 serves next, and so on
    /// round the six.
    public var offsetFromServer: Int { rawValue - 1 }

    /// The positions in the order they are drawn: front row left to right, then back row.
    public static let drawingOrder: [CourtPosition] = [
        .leftFront, .middleFront, .rightFront,
        .leftBack, .middleBack, .service,
    ]
}

/// One box on the court.
public struct CourtSlot: Equatable, Sendable {
    public var position: CourtPosition

    /// Who is standing here, or nil when the position is empty — a short bench, or a
    /// player removed from the roster mid-match. An empty position is shown as empty.
    public var playerId: String?

    public var number: String?

    /// Nil when this player has not served: they have no percentage, and reporting 0%
    /// would say they served and missed.
    public var inPercentage: Double?

    /// Nil where points were never recorded at all.
    public var points: Int?

    public var isServing: Bool
    public var isOnDeck: Bool

    public init(
        position: CourtPosition,
        playerId: String? = nil,
        number: String? = nil,
        inPercentage: Double? = nil,
        points: Int? = nil,
        isServing: Bool = false,
        isOnDeck: Bool = false
    ) {
        self.position = position
        self.playerId = playerId
        self.number = number
        self.inPercentage = inPercentage
        self.points = points
        self.isServing = isServing
        self.isOnDeck = isOnDeck
    }
}

/// The six boxes, and what the figures in them cover.
public struct CourtView: Equatable, Sendable {
    /// In drawing order: front row left to right, then back row.
    public var slots: [CourtSlot]

    /// The lineup index standing in the service corner.
    public var servingPosition: Int

    /// Nil when there is no order to advance, in which case the app must say it cannot
    /// name the next server rather than presenting one.
    public var onDeckPlayerId: String?

    /// What the figures cover, in words the coach can read: "Match 2", say.
    public var scopeLabel: String

    public init(
        slots: [CourtSlot],
        servingPosition: Int,
        onDeckPlayerId: String?,
        scopeLabel: String
    ) {
        self.slots = slots
        self.servingPosition = servingPosition
        self.onDeckPlayerId = onDeckPlayerId
        self.scopeLabel = scopeLabel
    }

    /// True when there is no lineup, so nobody can be named as next.
    public var hasOrder: Bool { onDeckPlayerId != nil }
}

/// The lineup index standing at a court position, wrapping round the order.
public func lineupIndex(servingPosition: Int, offset: Int) -> Int {
    let size = lineupSize
    return ((servingPosition + offset) % size + size) % size
}

extension AppState {
    /// The court as it stands in the match in progress, with figures for that match.
    ///
    /// Returns nil when there is no match to draw — the watch then says so rather than
    /// showing an empty court that looks like a court with nobody on it.
    public func courtView(scopeLabel: String? = nil) -> CourtView? {
        guard let match = currentMatch else { return nil }
        return courtView(of: match, scopeLabel: scopeLabel ?? "Match \(match.index + 1)")
    }

    /// The court for a given match.
    public func courtView(of match: Match, scopeLabel: String) -> CourtView {
        let servingPosition = serviceCornerPosition(of: match)
        let statistics = match.statistics
        let lineup = match.lineup

        let slots = CourtPosition.drawingOrder.map { position -> CourtSlot in
            let index = lineupIndex(servingPosition: servingPosition, offset: position.offsetFromServer)
            let playerId = lineup.flatMap { $0.indices.contains(index) ? $0[index] : nil }

            guard let playerId else { return CourtSlot(position: position) }

            let figures = statistics[playerId]
            return CourtSlot(
                position: position,
                playerId: playerId,
                number: rosterEntry(id: playerId)?.number,
                inPercentage: figures?.inPercentage,
                // Nil rather than zero: a player who has not served has earned no points
                // that were recorded, which is not the same as having earned none.
                points: figures?.points,
                isServing: position == .service,
                isOnDeck: position == .rightFront
            )
        }

        return CourtView(
            slots: slots,
            servingPosition: servingPosition,
            onDeckPlayerId: onDeckPlayerId(of: match, servingPosition: servingPosition),
            scopeLabel: scopeLabel
        )
    }

    /// Which lineup position is standing in the service corner.
    ///
    /// The open turn's own position comes first: it is where the ball actually is,
    /// including when a server out of the order took the position that was due.
    private func serviceCornerPosition(of match: Match) -> Int {
        if let position = match.openTurn?.lineupPosition { return position }
        return match.nextRotationPosition ?? 0
    }

    private func onDeckPlayerId(of match: Match, servingPosition: Int) -> String? {
        guard let lineup = match.lineup else { return nil }
        let index = lineupIndex(servingPosition: servingPosition, offset: 1)
        return lineup.indices.contains(index) ? lineup[index] : nil
    }
}

File: src/ui/components/court.js
// The six on court, laid out as the court itself: net at the top, service position at the
// bottom right, and the rotation running clockwise through the other five.
//
// The flat grid it replaces was a list of numbers in roster order, which asked the operator
// to hold the serving order in their head while watching the game. Here the arrangement IS
// the order: whoever is standing in the service corner has the ball, and the player who
// takes it next is the one directly above them, exactly as on the floor.
//
// Everyone else is a substitute, and is shown as one -- off the court, marked as bench.
import { chip } from './chip.js'
import { LINEUP_SIZE } from '../../domain/events.js'

// Court positions in the order they are drawn: front row left to right, then back row.
// The number is how far along the serving order that position stands from the server, so
// the whole court re-lays itself around whoever is serving now.
const LAYOUT = [
  { offset: 3, label: '4' },
  { offset: 2, label: '3' },
  { offset: 1, label: '2' },
  { offset: 4, label: '5' },
  { offset: 5, label: '6' },
  { offset: 0, label: '1' },
]

/**
 * The court and the bench.
 *
 * `servingPosition` is the lineup index standing in the service corner. `stateFor` marks a
 * chip exactly as it does anywhere else, so a serving or half-substituted player looks the
 * same here as in the picker it replaces.
 */
export function courtView(roster, lineup, options = {}) {
  const servingPosition = options.servingPosition ?? 0
  const stateFor = options.stateFor ?? (() => '')
  const action = options.action ?? 'select-server'

  const cells = LAYOUT
    .map((slot) => courtCell(roster, lineup, slot, servingPosition, stateFor, action))
    .join('')

  return `
    <div class="court">
      <div class="court-net" aria-hidden="true"></div>
      <div class="court-grid">${cells}</div>
    </div>
    ${benchRow(roster, lineup, stateFor, action)}`
}

/** The lineup index standing at a court position, wrapping round the order. */
export function positionAt(servingPosition, offset) {
  return (servingPosition + offset) % LINEUP_SIZE
}

function courtCell(roster, lineup, slot, servingPosition, stateFor, action) {
  const index = positionAt(servingPosition, slot.offset)
  const player = roster.find((each) => each.id === lineup[index])
  const classes = ['court-cell', slot.offset === 0 ? 'is-service' : '', slot.offset === 1 ? 'is-next' : '']
    .filter(Boolean).join(' ')

  // A slot can empty when a player leaves the roster mid-match. It is still a position in
  // the order, and drawing it says so -- silently closing the gap would misreport who
  // serves next.
  const body = player
    ? chip(player, { action, state: stateFor(player) })
    : '<div class="court-empty" aria-label="Nobody in this position">—</div>'

  return `
    <div class="${classes}">
      <span class="court-pos">${slot.label}</span>
      ${body}
      ${slot.offset === 0 ? '<span class="court-tag">serving</span>' : ''}
      ${slot.offset === 1 ? '<span class="court-tag court-tag-next">next</span>' : ''}
    </div>`
}

/**
 * Everyone not on the court. They are the other half of a substitution -- tapped first,
 * then the player they replace -- so they are kept plainly apart from the six who are on.
 */
function benchRow(roster, lineup, stateFor, action) {
  const bench = roster.filter((player) => !lineup.includes(player.id))
  if (bench.length === 0) return ''

  const chips = bench
    .map((player) => chip(player, { action, state: `is-bench ${stateFor(player)}`.trim() }))
    .join('')

  return `
    <div class="bench">
      <div class="bench-label">Bench</div>
      <div class="bench-chips">${chips}</div>
    </div>`
}
